runtime/drive_handler.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	firecracker "github.com/firecracker-microvm/firecracker-go-sdk"
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+)
+
+const (
+	minFCDriveImageBytes = 512
+)
+
+var (
+	// ErrStubDrivesExhausted occurs when there are no more stub drives left to
+	// use. This can happen by calling PatchStubDrive greater than the stub drive
+	// list.
+	ErrStubDrivesExhausted = fmt.Errorf("There are no remaining stub drives to use")
+
+	// ErrDriveIDNil should never happen, but we safe guard against nil dereferencing
+	ErrDriveIDNil = fmt.Errorf("DriveID of current drive is nil")
+
+	// ErrVMNotFound means that a VM with the given ID not found
+	ErrVMNotFound = errors.New("vm not found")
+)
+
+// stubDriveHandler is used to create a pool of stubbed file. This will allow
```

Comment:
Considering we're going to have one runtime process per VM, I don't think this pool works.  In particular, the `stubDriveHandler` struct is a per-process construct the way we're using it, which means the locks and map are also per-process and not shared.

I'm also not really sure there's a ton of value in sharing these stubs between VMs.  They're very small files, and we can have a natural garbage-collection point in cleaning them up when a VM exits if they're not shared.  It seems simpler to me to have separate stub files per VM and if we start to run into problems, we can always come back and add the more complex sharing construct.

Reply:
Can you update the comment here if you're not using a pool?